protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,853 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::protocol::{AutoNatCodec, AutoNatProtocol, DialRequest, DialResponse, ResponseError};
+use futures::FutureExt;
+use futures_timer::Delay;
+use instant::Instant;
+use libp2p_core::{
+    connection::{ConnectionId, ListenerId},
+    multiaddr::Protocol,
+    ConnectedPoint, Multiaddr, PeerId,
+};
+use libp2p_request_response::{
+    handler::RequestResponseHandlerEvent, ProtocolSupport, RequestId, RequestResponse,
+    RequestResponseConfig, RequestResponseEvent, RequestResponseMessage, ResponseChannel,
+};
+use libp2p_swarm::{
+    dial_opts::{DialOpts, PeerCondition},
+    AddressScore, DialError, IntoProtocolsHandler, NetworkBehaviour, NetworkBehaviourAction,
+    PollParameters,
+};
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    iter,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+/// Config for the [`Behaviour`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Config {
+    /// Timeout for requests.
+    pub timeout: Duration,
+
+    // == Client Config
```

Comment:
```suggestion
    /// Client Config
    
```

Nit pick: Do we do this (`==`) anywhere else? Why not just the above? As far as I can tell Rust fmt does not remove the newlines.

Reply:
The newlines are on purpose. The `//== Client Config` and `//== Server Config` comments are meant as mental "subheader" for the configs that are listed below them. I changed them now to just `// Client Config` / `// Server Config`, but I am also fine with completely removing them.